addons/payment/tests/test_flows.py

```diff
@@ -347,3 +326,44 @@ def test_transaction_wrong_token(self):
         self.assertIn(
             "odoo.exceptions.ValidationError: The access token is invalid.",
             response.text)
+
+    @mute_logger('odoo.addons.payment.models.payment_transaction')
+    def _test_max_one_payment_request_from_portal(self, flow):
+
+        payment_option_id = self.create_token().id if flow == 'token' else self.acquirer.id
+        data = {
+            'amount': self.amount,
+            'currency_id': self.currency.id,
+            'partner_id': self.partner.id,
+            'access_token': self._generate_test_access_token(
+                self.partner.id, self.amount, self.currency.id
+            ),
+            'payment_option_id': payment_option_id,
+            'reference_prefix': 'test',
+            'tokenization_requested': True,
+            'landing_route': 'Test',
+            'is_validation': False,
+            'invoice_id': self.invoice.id,
+            'flow': flow,
+        }
+
+        with patch(
+            'odoo.addons.payment.models.payment_transaction.PaymentTransaction.'
+            '_send_payment_request'
+        ) as patched:
+
+            response = self.portal_transaction(**data)
+            self.assertEqual(response.status_code, 200)
+
+            if flow == 'token':
+                patched.assert_called_once()
+            else:
+                patched.assert_not_called()
```

Comment:
I'm not too fan of 'fourre-tout' tests whose assertions vary depending on the input. Wouldn't be easier/clearer to debug if you moved `data` as a constant in `http_common.py` and split `test_max_one_payment_request_from_portal` into as many tests as you like? For example:
- `test_direct_payment_triggers_no_payment_request`
- `test_payment_with_redirect_triggers_no_payment_request`
- `test_payment_by_token_triggers_exactly_one_payment_request`

Reply:
Been there, done that, I felt like there were just a bit too many repetitions. However, I might have found a middle ground that I like. I could go back to the 3 tests instead if you prefer. =]